.github/workflows/publish-sql-schema.yml

```diff
@@ -24,8 +30,40 @@ jobs:
       run: |
         VERSION=${GITHUB_REF/refs\/tags\//}
         echo "MINOR_VERSION=$(echo ${VERSION} | awk -F'.' '{if (NF != 3) {exit 1}; printf "%s.%s", $1, $2}')" >> $GITHUB_OUTPUT
+    - name: Check validity of migrations
+      uses: "./.github/actions/validate-migrations"
+    - name: Check that existing migrations have not changed
+      if: '!inputs.force'
+      run: |
+        # Overwite the migrations on the local ref so that we can use `git diff` to compare.
+        rm -f db/*.sql
+        trap "git reset --hard" EXIT
+        gcloud alpha storage cp --recursive "gs://janus-artifacts-sql-schemas/$MINOR_VERSION/db/*.sql" db/
+
+        git add db/*.sql
+        DIFF=$(git diff --name-status HEAD db/)
+        printf "%s\n" "$DIFF"
+
+        # The meaning of added and removed files are reversed when comparing this
+        # way. If a file comes up deleted, that means that the local git ref has
+        # introduced a new file that is not present in the remote. This is permissible.
+        # If a file comes up added, that means that the local ref has removed an
+        # existing migration which was readded by the overwrite step. This is not
+        # permitted.
+        if [ -n "$DIFF" ] && grep -qPv '^D\b' <<<"$DIFF"; then
+          echo "fatal: migrations cannot be modified or removed, only added."
+          exit 1
+        fi
```

Comment:
I don't really like this approach, it's quite hacky. I'm open to other ideas.

The goal is as so:
- Differences between any existing files on either side should be rejected.
- Files missing on the remote but present in the local should be permitted. (the local is adding a new migration)
- Files present on the remote but absent from the local should be rejected. (the local has deleted a migration)

`diff -r` is not sufficient, because it doesn't differentiate between local/remote differences in a machine-parseable way. I looked at `rclone` to see if it could instead entirely replace `gcloud alpha storage`, but I don't see the right set of flags to make it work as described. We also could just leave out this step and let `gcloud alpha storage cp --no-clobber` do the work, but this has poor UX as it will just bail out with an ambiguous HTTP error (and does not cover for the 3rd case).

Reply:
I'm inclined to agree that parsing the output of `git diff --name-status` is dicey:

- I find regexes hard to read (though this one is simple enough that even I was able the understand it, so a short comment explaining what it does would suffice)
- This would break if the format of `diff --name-status` ever changes (I'm not sure if `git` makes guarantees about the stability of its output format)

In any case, while poking around the `git-diff` man page, I think I found something that I think takes the risk out of this and is more self documenting: [`--diff-filter`](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---diff-filterACDMRTUXB82308203). If we did `git diff --diff-filter=DM`, then only files that are `D`eleted or `M`odified are included in the diff, and all you have to do is check if the output is empty. I suppose we could do `--diff-filter=CDMRTUXB` to catch everything _except_ added files.

All that said, you might mean that doing `rm db/*` and replacing it with the contents of the GCS bucket is hacky. Perhaps slightly better is [`git diff --no-index`](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt-emgitdiffemltoptionsgt--no-index--ltpathgtltpathgt). So you could:

```
git diff --no-index \
    --diff-filter=CDMRTUXB \ # Allow [A]dditions but nothing else
    $BUCKET/db \ # contents of the GCS bucket
    $CHECKOUT/db # contents of the git checkout
```

Then if anything has changed between `$BUCKET/db` and `$CHECKOUT/db` that isn't an addition, the output will be non-empty and you can print the diff and fail the action.

Reply:
That's excellent! The result looks great!
```yaml
    - name: Check that existing migrations have not changed
      if: '!inputs.force'
      run: |
        TEMP=$(mktemp -d)
        gcloud alpha storage cp --recursive \
          "gs://janus-artifacts-sql-schemas/${{ steps.get_version.outputs.minor_version }}/db"\
          "$TEMP"
        if ! git diff --no-index --name-status --diff-filter=CDMRTUXB "$TEMP/db" "./db"; then
          echo "fatal: migrations cannot be modified or removed, only added"
          exit 1
        fi
```
Plus a bonus fix for missing string interpolation (was using `$MINOR_VERSION` to test the script in isolation).

I'll carry it back to https://github.com/divviup/janus/blob/main/.github/workflows/ci-migrations.yml#L19 as well.